### PR TITLE
EZP-29586: Content copied by another user is still owned by original creator

### PR DESCRIPTION
**Pull request by ViniTou**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29586](https://jira.ez.no/browse/EZP-29586)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `6.7`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

New owner is set for very copied content via `copy` or `copySubtree` method.

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by ViniTou** [comment]
@andrerom 
@alongosz 

rebased, but this probably should be moved somewhere else
https://github.com/ezsystems/ezpublish-kernel/pull/2447/files#diff-6a78cd6b49cc43af24f88b53b9ba100e

File: eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php [hunk 1]
```diff
@@ -438,15 +457,21 @@ protected function extractFieldValueFromRow(array $row, $type)
...
     {
         $struct = new CreateStruct();
         $struct->name = $content->versionInfo->names;
         $struct->typeId = $content->versionInfo->contentInfo->contentTypeId;
         $struct->sectionId = $content->versionInfo->contentInfo->sectionId;
-        $struct->ownerId = $content->versionInfo->contentInfo->ownerId;
+        $struct->ownerId =
+            $this->permissionResolver
+                ? $this->permissionResolver->getCurrentUserReference()->getUserId()
+                : $content->versionInfo->contentInfo->ownerId;
```
[the remark this answers is not in the document]
Comment (ViniTou) [reply eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php]: So set it over here?
`\eZ\Publish\Core\Persistence\Legacy\Content\Handler::copy`
that's not gonna change anything around circular reference problem but I don't know if that was your point?
Or you mean just setting `$ownerId` as an optional argument cuz I may had misunderstood you at first :D

[the remark this answers is not in the document]
Comment (andrerom) [reply eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php]: > Or you mean just setting $ownerId as an optional argument cuz I may had misunderstood you at first :D

this is what I suggested first.
aka `createCreateStructFromContent(Content $content, $newOwnerId = null)`

Which I think can make sense here.

An alternative, if that does not fly for some reason,, would be that callers of this do:
```php
$struct = $obj->createCreateStructFromContent($content);
$struct->ownerId = $currentUserId;
```

Both avoids the injection done here..

[the remark this answers is not in the document]
Comment (ViniTou) [reply eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php]: There is only one problem I could see with both of those solutions - if someone miss setting ownerId either in parameter (if we would go with optional one) to method or after using `createCreateStructFromContent` we gonna end with content without owner, right?

@andrerom

[the remark this answers is not in the document]
Comment (adamwojs) [reply eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php]: +1 for solution suggested by @andrerom

In summary solution might looks as follows:

1. Add `$newOwnerId` as optional parameter to the `\eZ\Publish\Core\Persistence\Legacy\Content\Handler::copy` 
2. In \eZ\Publish\Core\Repository\ContentService::createContent get the current user id form PermissionResolver and pass it into `\eZ\Publish\Core\Persistence\Legacy\Content\Handler::copy` 
 method call here  https://github.com/ezsystem/ezpublish-kernel/blob/c31c285bc884392b62c7afd9b23a38b4cdfb8ff3/eZ/Publish/Core/Repository/ContentService.php#L1662-L1665
3. If the `$newOwnerId`is not null then assign it to the `$createStruct->ownerId` (in this way by default will be used current behaviour)

Optionally you can also add `$newOwnerId` as parameter of `\eZ\Publish\Core\Persistence\Legacy\Content\Mapper::createCreateStructFrom Content` method.

[the remark this answers is not in the document]
Comment (andrerom) [reply eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php]: > setting ownerId either in parameter (if we would go with optional one) to method or after using createCreateStructFromContent we gonna end with content without owner, right?

as @adamwojs says, ownerId arg will be optional, so when not set current logic of `$content->versionInfo->contentInfo->ownerId` will set owner like before your PR.

File: eZ/Publish/API/Repository/Tests/ContentServiceTest.php [hunk 2]
```diff
@@ -2838,6 +2838,59 @@ public function testCopyContent()
...
+        $repository = $this->getRepository();
+
+        $contentService = $repository->getContentService();
+        $locationService = $repository->getLocationService();
+
+        $newOwner = $this->createUser('new_owner', 'foo', 'bar');
+        /* BEGIN: Use Case */
+        /** @var \eZ\Publish\API\Repository\Values\Content\Content $contentVersion2 */
+        $contentVersion2 = $this->createContentDraftVersion1(
+            56,
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/ContentServiceTest.php]: You're relying too much on generated ids.

File: eZ/Publish/API/Repository/Tests/ContentServiceTest.php [hunk 3]
```diff
@@ -2838,6 +2838,59 @@ public function testCopyContent()
...
+            $targetLocationCreate
+        );
+        /* END: Use Case */
+
+        $this->assertEquals(
+            $newOwner->id,
+            $contentVersion2->contentInfo->ownerId
+        );
+        $this->assertEquals(
+            14,
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/ContentServiceTest.php]: Here as well.

File: eZ/Publish/Core/Persistence/Cache/LocationHandler.php [hunk 4]
```diff
@@ -122,11 +122,11 @@ public function loadByRemoteId($remoteId)
     /**
      * @see \eZ\Publish\SPI\Persistence\Content\Location\Handler::copySubtree
      */
-    public function copySubtree($sourceId, $destinationParentId)
+    public function copySubtree($sourceId, $destinationParentId, $newOwnerId = null)
     {
         $this->logger->logCall(__METHOD__, array('source' => $sourceId, 'destination' => $destinationParentId));
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Cache/LocationHandler.php]: Usually it is a good idea to add new argument to the log here.

**Comment by andrerom** [comment]
@micszo This is very ripe for QA btw :)

**Comment by micszo** [comment]
And it is next on the list. ;)

**Comment by micszo** [comment]
_Details_ and _Versions_ tabs are covered. What about Author on _View_ tab? 
This is with patch after being copied by Administrator.
screen shot 2018-10-05 at 13 54 22

**Comment by andrerom** [comment]
Author _Field_ won't be affected, this is "Content". What is affected here is the Content->creator _property_, and maybe also Content->modifier @ViniTou ?

**Comment by ViniTou** [comment]
`content->ownerId` to be precise and prolly `content->versionInfo->ownerId` for version contributor.
Author is fieldType used in specific contentType, don't know if this is part of this scope - and if it is it completely missed me.

**Comment by andrerom** [comment]
My take on Expectations here:
- `ContentInfo->owner` should be updated
- `VersionInfo->creator` stays what it was so version info stays true to original source
- Author field is untouched, it's content and should not at all be changed

I'm unsure about `ContentInfo->modifier`, it can stay what it was or be changed, both can be argued to be correct behavior.
WDYT @alongosz ?

**Comment by micszo** [comment]
@andrerom To sum up. 
On Details tab Creator and Last contributor should be updated and on Versions tab Author? 
That's how it works now with patch.

**Comment by andrerom** [comment]
> On Details tab Creator and Last contributor should be updated and on Versions tab Author?
That's how it works now with patch.

You'll need to check that with someone else, I'm a bit unsure which api properties are used for what in UI without having to check quite some layers of UI code ;)

!! But I'm a bit surprised user name in Versions tab is changed to be honest, so maybe it's a reason to hold this one back a bit to take a deeper look next week when we are not preoccupied with prepping for release.

**Comment by ViniTou** [comment]
@micszo 
can you confirm that those same fields are also changed in 2.X?

**Comment by micszo** [comment]
@ViniTou not yet, so far have checked 1.7.8 and 1.13.4. Working on it.

**Comment by andrerom** [comment]
> @ViniTou not yet, so far have checked 1.7.8 and 1.13.4. Working on it.

good point, check 2.x. I think 1.x might be doing things in this regards quite a lot different.

**Comment by SylvainGuittard** [comment]
> On Details tab Creator and Last contributor should be updated and on Versions tab Author?

Yes, that is what happens in eZ Publish. The reason is that the user should be able to edit the content item and also restore a previous version. Especially when you have a policy like `content/edit` with limitation owner `self`.
Hope this helps

**Comment by ViniTou** [comment]
So if QA pass on all versions in that same behavior as in 1.7.8 and 1.13.4, all is ok?

**Comment by SylvainGuittard** [comment]
@ViniTou maybe it's less problematic on 1.x because the user permission regarding `content/edit` with limitations are perfectly handled.
But I will recommend to have the same behavior on all of our versions

**Comment by micszo** [comment]
@ViniTou on 2.2.3 with patch also all three values are changed on Details and Versions tabs.

**Comment by micszo** [comment]
Reading again into @SylvainGuittard's comments I think it works as required, same way on all versions (1.7.8, 1.13.4, 2.2.3).

**Comment by lserwatka** [comment]
@ViniTou could you merge it up to 6.13, 7.2 and master (still 7.3.x)

**Pull request merged**